src/basic/macro.h

```diff
@@ -318,10 +322,14 @@ static inline int __coverity_check_and_return__(int condition) {
                         *p = func(*p);                          \
         }
 
-/* When func() doesn't return the appropriate type, set variable to empty afterwards */
+/* When func() doesn't return the appropriate type, set variable to empty afterwards.
+ * The func() may be provided by a dynamically loaded shared library, hence add an assertion. */
 #define DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(type, func, empty)     \
         static inline void func##p(type *p) {                   \
                 if (*p != (empty)) {                            \
+                        DISABLE_WARNING_ADDRESS;                \
+                        assert(func);                           \
+                        REENABLE_WARNING;                       \
```

Comment:
instead of fiddling with the warning levels, couldn't we just do something like this:

```c
typeof(func) _func = func;
assert(_func);
_func(*p);
```

i.e. take a function pointer first, and check that.

Reply:
Unfortunately, it does not work:
```
In file included from ../src/libsystemd/sd-bus/bus-message.c:14:
../src/basic/fd-util.h: In function ‘pclosep’:
../src/basic/fd-util.h:48:1: error: function ‘_func’ is initialized like a variable
   48 | DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(FILE*, pclose, NULL);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/basic/macro.h:13,
                 from ../src/basic/alloc-util.h:10,
                 from ../src/libsystemd/sd-bus/bus-message.c:9:
../src/basic/macro.h:173:32: warning: the address of ‘_func’ will always evaluate as ‘true’ [-Waddress]
  173 |                 if (_unlikely_(!(expr)))                                \
      |                                ^
../src/fundamental/macro-fundamental.h:32:44: note: in definition of macro ‘_unlikely_’
   32 | #define _unlikely_(x) (__builtin_expect(!!(x), 0))
      |                                            ^
../src/basic/macro.h:190:22: note: in expansion of macro ‘assert_message_se’
  190 | #define assert(expr) assert_message_se(expr, #expr)
      |                      ^~~~~~~~~~~~~~~~~
../src/basic/macro.h:327:25: note: in expansion of macro ‘assert’
  327 |                         assert(_func);                          \
      |                         ^~~~~~
../src/basic/fd-util.h:48:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC_FULL’
   48 | DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(FILE*, pclose, NULL);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/basic/fd-util.h: In function ‘closedirp’:
../src/basic/fd-util.h:49:1: error: function ‘_func’ is initialized like a variable
   49 | DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(DIR*, closedir, NULL);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/basic/macro.h:173:32: warning: the address of ‘_func’ will always evaluate as ‘true’ [-Waddress]
  173 |                 if (_unlikely_(!(expr)))                                \
      |                                ^
../src/fundamental/macro-fundamental.h:32:44: note: in definition of macro ‘_unlikely_’
   32 | #define _unlikely_(x) (__builtin_expect(!!(x), 0))
      |                                            ^
../src/basic/macro.h:190:22: note: in expansion of macro ‘assert_message_se’
  190 | #define assert(expr) assert_message_se(expr, #expr)
      |                      ^~~~~~~~~~~~~~~~~
../src/basic/macro.h:327:25: note: in expansion of macro ‘assert’
  327 |                         assert(_func);                          \
      |                         ^~~~~~
../src/basic/fd-util.h:49:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC_FULL’
   49 | DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(DIR*, closedir, NULL);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```